### ISSUE TITLE
Proposal: classList casts values to Boolean

### ISSUE THREAD

**Issue by edemaine** [solidjs/solid#989]
## Proposal

Currently, `classList={{foo: foo(), bar: bar()}}` compiles to
```js
effect(_p$ => {
  const _v$ = foo(),
        _v$2 = bar();

  _v$ !== _p$._v$ && _el$.classList.toggle("foo", _p$._v$ = _v$);
  _v$2 !== _p$._v$2 && _el$.classList.toggle("bar", _p$._v$2 = _v$2);
  return _p$;
}, {
  _v$: undefined,
  _v$2: undefined
});
```
I propose that the leading lines compile to this:
```js
  const _v$ = Boolean(foo()),
        _v$2 = Boolean(bar());
```
or if we're optimizing code length:
```js
  const _v$ = !!foo(),
        _v$2 = !!bar();
```
The `classList` helper would need similar modification.

## Motivation

* Efficiency: If `signal()` changes from `1` to `2`, we'll still call `toggle` even though it didn't change its Boolean state.
* `undefined` has particularly weird behavior, because `toggle` treats it as an absent argument, so it actually toggles the state of the class. [Here's a Playground example](https://playground.solidjs.com/?hash=304148137&version=1.4.1) which abuses this to cause the class to alternate on/off when a signal changes, which seems gross. A value that's always `undefined` doesn't trigger this behavior, because of the duplicate detection, but if you start at e.g. `false` and then become `undefined`, you'll enter weird territory.

## More

I also notice that a single class in `classList` doesn't get the "modified" detection: `classList={{foo: foo()}}` compiles to:
```js
effect(() => _el$.classList.toggle("foo", foo()));
```
This has especially bad behavior with `undefined`: if `foo()` is even initially `undefined`, it *turns the class on*. So we want to do the Boolean casting here for consistency:
```js
effect(() => _el$.classList.toggle("foo", Boolean(foo())));
```
Should this case also have duplicate detection to avoid calling `toggle`, or is this not a big enough deal? (Mainly I'm surprised that this compiles differently.)

[as [discussed on Discord](https://discord.com/channels/722131463138705510/751355413701591120/976119358222905444)]

**Comment by mosheduminer** [comment]
How might this affect
```
classList={{
  [props.class!]: !!props.class,
}}
```
? This is something I do all the time.

**Comment by edemaine** [comment]
> the effect doesn't care how those are manipulated by the function, it will just fire whenever they change regardless of whether they are wrapped in `Boolean` or not

Yes, the effect will fire in any case (unless we wrapped in memos like `Show` does, but that's probably overkill). The point is to stop from hitting the DOM if we can avoid it. `toggle` will need to check whether the class exists, which costs time. Admittedly this is a pretty small DOM access, but it's still an avoidable one.

> [props.class!]: !!props.class,

With my proposal it would end up computed `!!!!props.class`, which is very loud 🙂 but equivalent.  We could do some compile-time checks for when the value is definitely a Boolean (e.g. already `!!` or `Boolean(...)` at the outermost), and avoid adding excess `!`s then.

**Comment by edemaine** [comment]
I wanted to double-check the behavior here so that I can add it to docs, but I haven't been able to see `!!` get produced in the output, using `solid-js@1.4.2` and `vite-plugin-solid@2.2.6`. Am I missing something? Maybe the Vite plugin needs to be updated?

**Comment by titoBouzout** [comment]
As far as I know classList is meant to be removed.... so.....what's the alternative? improve class to work as classList?

**Comment by mosheduminer** [comment]
What do you mean? Are you thinking of the removal of className? @titoBouzout

**Comment by titoBouzout** [comment]
It seems to me (or just my perception) that classList is going to be removed, because well..... there's the fight of class vs classList  vs spread, which seems to be very hard on SSR as everything just clashes.

**Comment by LXSMNSYC** [comment]
@titoBouzout there was no mention of it? unless you can show us some discussion about it

**Comment by ryansolid** [comment]
In discord. We've talked about how classList is very awkward. There are a few different options though.
